### ISSUE TITLE
Better search algorithm

### ISSUE THREAD

**Issue by Timu57** [rakibtg/SleekDB#37]
Hi,

first of all great work and project.

It looks like the build in similar_text() function is not perfect.

I just made some little tests, but it looks like the returned percentages are not always what I expected.

**It would be awesome if we can add some his string normalization**:
https://stackoverflow.com/questions/35923029/similar-text-percentage-in-php

```
I have implemented several algorithms that will search for duplicates and they can be quite similar.

The approach I am usually using is the following:

    normalize the strings
    use a comparison algorithm (e.g. similar_text, levenshtein, etc.)

It appears to me that in implementing step 1) you will be able to improve your results drastically.

Example of normalization algorithm (I use "Sponsors back away from Sharapova after failed drug test" for the details):

1) lowercase the string

-> "sponsors back away from sharapova after failed drug test"

2) explode string in words

-> [sponsors, back, away, from, sharapova, after, failed, drug, test]

3) remove noisy words (like propositions, e.g. in, for, that, this, etc.). This step can be customized to your needs

-> [sponsors, sharapova, failed, drug, test]

4) sort the array alphabetically (optional, but this can help implementing the algorithm...)

-> [drug, failed, sharapova, sponsors, test]

Applying the very same algorithm to your other string, you would obtain:

[australian, drugs, failed, maria, open, sharapova, test]

This will help you elaborate a clever algorithm. For example:

    for each word in the first string, search the highest similarity in the words of the second string
    accumulate the highest similarity
    divide the accumulated similarity by the number of words

```

**And/Or add some math from MySQLs Fulltext-Search score algorithm**:
https://dba.stackexchange.com/questions/111738/mysql-match-relavancy-score
> Relevance is computed based on the number of words in the row, the number of unique words in that row, the total number of words in the collection, and the number of documents (rows) that contain a particular word.

## Add custom ranking key property to result

That ranking key can then be used in orderBy to sort the result.

**Comment by rakibtg** [comment]
@Timu57 This would be a great improvement, I will have to look at it in this weekend and hoping that it would be an easy implementation.

**Comment by nickbe** [comment]
Any news regarding rated queries?

**Comment by rakibtg** [comment]
@nickbe No updates yet, I think I will work on it after the next release.

**Comment by Timu57** [comment]
Implementing fulltext search in boolean mode like described here:
https://dev.mysql.com/doc/refman/5.6/en/fulltext-boolean.html

Does someone have an idea what could be the best solution to implement the different operators `-`, `+`, `~`, `<`, `>`...
Or should we ignore the operators for now?

**Comment by Timu57** [comment]
https://github.com/rakibtg/SleekDB/pull/138#issue-575485487
Description of new implementation.

**Comment by Timu57** [comment]
Implemented with version 2.7.
